### ISSUE TITLE
previous window's content shown as "ghost" for short moment

### ISSUE THREAD

**Issue by m1027** [labwc/labwc#870]
Hi,

I open firefox via keyboard and with a blank window by default.

When I open firefox, then open an arbitrary site (to have some content), close it and restart firefox, I see the previous site in the window for approx 0.25 sec.

I am am on this labwc commit from master:

commit e44d56567284c5860c1c3148c09628b50ce477db (HEAD -> master)
Author: Consolatis <35009135+Consolatis@users.noreply.github.com>
Date:   Tue Mar 28 19:23:49 2023 +0200

    src/cursor.c: make sure the cursor is visible after startup

Let me know when you need more info to reproduce that.

**Comment by Consolatis** [comment]
I don't remember any changes during the last few month or so that could explain that.

Some questions to try to pinpoint it to something:

Does it happen with both, `MOZ_ENABLE_WAYLAND=1` and without that env var set?
You can also verify via `A-Tab` if firefox is running as wayland native (`[xdg-shell]` in first column) or via xwayland (`[xwayland]` in first column).

I assume you are using a stock wlroots version from the 0.16 branch (either as system installation or as meson subproject), e.g. no wlroots-eglstream or other fork.

Did this start to happen just recently and if yes, what was the last version where it didn't happen?

Nvidia GPU drivers?

Finally, do you see a similar behavior with any other application?

**Comment by m1027** [comment]
> Does it happen with both, `MOZ_ENABLE_WAYLAND=1` and without that env var set? You can also verify via `A-Tab` if firefox is running as wayland native (`[xdg-shell]` in first column) or via xwayland (`[xwayland]` in first column).

Hm, I am on pure wayland, no xwayland installed.

> I assume you are using a stock wlroots version from the 0.16 branch (either as system installation or as meson subproject), e.g. no wlroots-eglstream or other unofficial fork / patchset, same for xwayland if in use.

Exactly. Pure wlroots-0.16.2.

> Did this start to happen just recently and if yes, what was the last version where it didn't happen?

I am using labwc only recently. Always on the the commit mentioned above. BTW: I just pulled master and recompiled, same situation.

> Nvidia GPU drivers?

No. Intel, as directly integrated into mesa.

> Finally, do you see a similar behavior with any other application or is it just firefox (which could point to a generic firefox issue unrelated to labwc / wlroots).

Not really. And I wouldn't know how either. Usually, all apps draw something immediately (like libreoffice, gimp etc). So, I don't see this.

I wonder what kind of image buffer is responsable for this at all? Firefox is not running between the test, so I currently assume it is something not related to firefox. That's why I am asking here.

I could try to go back to wayfire for another test.

**Comment by m1027** [comment]
I cross-checked against wayfire: I don't have the issue there (I don't see a ghost buffer when terminating and starting firefox).

**Comment by johanmalm** [comment]
Might this be a swapchain / scanout issue relating to the scene-graph implementation?

Would you be able to test with this sway branch?
https://github.com/Nefsen402/sway/tree/scene-graph

**Comment by johanmalm** [comment]
If you run “labwc -d 2> log.txt” do you get any swapchain related messages?

**Comment by m1027** [comment]
Thanks @johanmalm. Before I come back on your tests:

I can narrow down the case as follows. Firefox shows its latest website content temporarily after closing and re-opening only in this case:
- The alacritty terminal has been opened before opening firefox.
- The window of alacritty has been put into fullscreen. In my case I do it by pressing A-F11, having this config: `<keybind key="A-F11"><action name="ToggleFullscreen"/></keybind>`.

If I do the same without putting alacritty into fullscreen or when using e.g. libreoffice fullscreened instead of alacritty, firefox does not show the previous "ghost" content. It also does not matter to wait some seconds before reopening firefox until all its background processes are entirely terminated.

I hope you can reproduce with this this extra information?

**Comment by m1027** [comment]
> If you run “labwc -d 2> log.txt” do you get any swapchain related messages?

Yes. Initially I get the following line 3 times: `00:00:00.399 [DEBUG] [render/swapchain.c:105] Allocating new swapchain buffer
`

And another hint: It is harder to detect but it seems that I see some temporary ghost content in firefox even when opening firefox for the first time (but still only when launched on top of a fullscreened alacritty window).

**Comment by Consolatis** [comment]
Just a theory, not verified at all:

This sounds like firefox is submitting an empty buffer (e.g. allocated but not actually written to). I am not sure who is responsible for cleaning the buffers (and if it should be done in the first place, seems pretty useless). Or if its just a firefox issue that it submits the buffer before actually rendering something into it during startup.

Depending on various circumstances the content of the allocated buffer may then contain parts of the earlier output or something completely different.

**Comment by m1027** [comment]
> With the fullscreen alacritty + Firefox it feels direct scanout related.
> 
> Do you get the same with Firefox starting in a non-maximised state?

Yes. Alacritty needs to be fullscreen but firefox can be non-maximized to trigger this issue.

And one more hint: If there is another firefox window already open (and kept open somewhere) then for newly added firefox instances this issue is *not* triggered. You need to close *all* instances of firefox and then reopen firefox to trigger the issue.

**Comment by m1027** [comment]
> Just a theory, not verified at all:
> 
> This sounds like firefox is submitting an empty buffer (e.g. allocated but not actually written to). I am not sure who is responsible for cleaning the buffers (and if it should be done in the first place, seems pretty useless). Or if its just a firefox issue that it submits the buffer during startup before actually rendering something into it.

Please note: I wasn't able to reproduce this issue when just launching wayfire WM instead of labwc.

IMHO: The whole thing is a bit more than just a  little flicker effect. There might be confidential stuff being brought up on screen unexpectedly at a time when even others my be watching.

**Comment by Consolatis** [comment]
Does it happen as well when starting labwc with software rendering, e.g. `WLR_RENDERER=pixman labwc` ?

**Comment by m1027** [comment]
> Does it happen as well when starting labwc with software rendering, e.g. `WLR_RENDERER=pixman labwc` ?

Display operations are getting very slow, but no, it doesn't happen then.

**Comment by Consolatis** [comment]
Just saw this, its possibly the same issue: https://github.com/swaywm/sway/pull/6844#issuecomment-1509344115
So starting labwc like  `WLR_SCENE_DISABLE_VISIBILITY=1 labwc` might help.

**Comment by m1027** [comment]
> So starting labwc like `WLR_SCENE_DISABLE_VISIBILITY=1 labwc` might help.

Confirmed.

Some more notes:

- With `WLR_SCENE_DISABLE_VISIBILITY=1 labwc` labwc feels a little less crisp when opening and closing windows. I assume this is a workaround and things should be fixed.
- Just another test: It is not necessarily the previous firefox content to show up as ghost content when reopening firefox. You can e.g. open libreoffice in between and you will see the libreoffice window contents in firefox for a short period.
- I refuse to believe ;-) that this is a firefox bug. Firstly, I don't see this in wayfire WM (for whatever reason). Secondly, IMHO it would be a huge security issue if it was left to newly starting apps to wipe their window buffers before showing it. What if they just don't (like maybe firefox)? Virtually any app could then read sensitive parts of the screen.

In case this is not a labwc issue, feel free to comment and close.

Thanks to everybody.

**Comment by Consolatis** [comment]
With the new hint it seems my earlier theory is wrong.

What seems to actually be the case is that there is a bug in firefox which, during startup, claims that a surface is opaque while in reality its translucent. This then causes a optimization pass in the render backend to show the previous part of the screen in that place. All wlroots based compositors that use the scene graph rendering backing are likely affected by this. Setting `WLR_SCENE_DISABLE_VISIBILITY=1` disables the optimization, thus also renders surfaces that are (or should) not be visible anyway, e.g. whatever is behind of the "fully opaque" firefox surface.

So while this indeed seems to be a bug in firefox, it won't allow it to read old content from its buffers but instead "just" causes the rendering backend to show part of earlier content in its place.

**Comment by johanmalm** [comment]
@Consolatis Thanks for finding the river and sway-scene-graph issues and also homing in on the optimization relating to `WLR_SCENE_DISABLE_VISIBILITY`. Be nice to get to the bottom of this one, but think I'll focus on some other issues for the minute when I get back into gear.

**Comment by xxhh1112** [comment]
eqeqeqw
